docs(migration): Added step-by-step guide for the Auth0 to Zitadel migration (#10118)
zitadel/zitadel: docs/docs/guides/migrate/sources/auth0-guide.md

## 5. Importing Users into ZITADEL

### 5.1. Obtain Access Token (or PAT) for API Access

To call the ZITADEL Admin API, you need to authenticate using a **Service User** with the `IAM_OWNER` Manager permissions.

There are two recommended authentication methods:

- **Client Credentials Flow**  
  [Learn how to authenticate with client credentials.](https://zitadel.com/docs/guides/integrate/service-users/client-credentials)

- **Personal Access Token (PAT)**  
  [Learn how to create and use a PAT.](https://zitadel.com/docs/guides/integrate/service-users/authenticate-service-users#personal-access-token)

**Reference:** [Service Users & API Authentication](https://zitadel.com/docs/guides/integrate/service-users/authenticate-service-users#authentication-methods)

---

### 5.2. Import Data with the ZITADEL API

- Use your **access token** or **PAT** to authenticate.
- Call the [Admin API – Import Data](https://zitadel.com/docs/apis/resources/admin/admin-service-import-data) endpoint, passing your generated JSON file.
- Verify that the users were imported successfully in the ZITADEL console.

**Import Endpoint:**

- `POST /admin/v1/import`
- `Authorization: Bearer <token>`

> - **Body:** Generated in step 4.2

Reviewer: I think we should looking into extending the zitadel tools to automatically upload.

I.e. have a compiled CLI that you can feed with the envs like instance, token and an import file.

Not important now but might be a good improvement.

Author: Yes, I absolutely agree. The current tool is just a file formatting tool, not much of a CLI tool. Another point to add as an enhancement request to take a deeper look after the workshop.

---
title: Migrating Users from Auth0 to ZITADEL (Including Password Hashes)
sidebar_label: Auth0 Migration Guide
---

## 1. Introduction

> **Why migrate from Auth0 to ZITADEL?**  

Reviewer: Having this question here feels a little weird since we are not answering it. 

We should either have an answer to this our link to some resources 😁

> This guide will walk you through the steps to migrate users from Auth0 to ZITADEL, including password hashes (with Auth0 support assistance), so users don't need to reset their passwords.

Reviewer: ```suggestion
This guide will walk you through the steps to migrate users from Auth0 to ZITADEL, including password hashes (which requires Auth0's support assistance), so users don't need to reset their passwords.
```

**What you'll learn with this guide**
- How to prepare your data from Auth0

> - Use of ZITADEL migration tooling

Reviewer: ```suggestion
- Use of the ZITADEL migration tooling
```

> - Performing the user import via API

Reviewer: ```suggestion
- Performing the user import via ZITADEL's API
```

- Troubleshooting and validating the migration

---

## 2. Prerequisites

### 2.1. Install Go

> The migration tool is written in Go. Download and install the latest version of Go from the [official Go website](https://go.dev/).

Reviewer: Would it not make more sense to have this as a compile binary?

Not urgent but would be a good QoL improvement

Author: Yes, that's a good point. I can create an enhancement request for the zitadel-tools after this workshop, would that work?

### 2.2. Create a ZITADEL Instance and Organization
You'll need a target organization in ZITADEL to import your users. You can create a new organization or use an existing one.  

> If you don't have one, you can [sign up for free here](https://zitadel.com).  

Reviewer: ```suggestion
If you don't have one, you can [sign up for free here](https://zitadel.com) to create a new ZITADEL for you.  
```

See: [Managing Organizations in ZITADEL](https://docs.zitadel.com/).

> > **Note:** Copy your Organization ID (Resource ID) for use in later steps.

Reviewer: ```suggestion
> **Note:** Copy your Organization ID (Resource ID) since you will use the id in the later steps.
```

> **Important:** Password hashes cannot be obtained in a self-service way.  
> You must open a **support ticket** with Auth0 and request a password hash export.  
> If approved, Auth0 will provide an export containing the password hashes.  

Reference: [Export hashed passwords from Auth0](https://auth0.com/docs/users/export-users/export-users-passwords)

---

## 4. Running the ZITADEL Migration Tool

### 4.1. Install the Migration Tool
Follow the installation instructions to set up the ZITADEL migration tool from [ZITADEL Tools](https://github.com/zitadel/zitadel-tools).

### 4.2. Generate Import JSON
Use the migration tool to convert the Auth0 export file into a ZITADEL-compatible JSON.

**Example workflow:**

Sample `passwords.ndjson`:
```json
{"_id":{"$oid":"emxdpVxozXeFb1HeEn5ThAK8"},"email_verified":true,"email":"tommie_krajcik85@hotmail.com","passwordHash":"$2b$10$d.GvZhGwTllA7OdAmsA75uGGzqr/mhdQoU88M3zD.fX3Vb8Rcf33.","password_set_date":{"$date":"2025-06-30T00:00:00.000Z"},"tenant":"test","connection":"Username-Password-Authentication","_tmp_is_unique":true}

> ```

Reviewer: Is this the sample from auth0 or zitadel? We should make this clear.

Author: Changed it to:
`Sample `passwords.ndjson` content, as obtained from the Auth0 Support team:`

# Migrating Users into ZITADEL

## Sample profiles.ndjson content

```json
{"user_id":"auth0|emxdpVxozXeFb1HeEn5ThAK8","email_verified":true,"name":"Tommie Krajcik","email":"tommie_krajcik85@hotmail.com"}

> ```

Reviewer: These headings look weird, also we should make clear for what system the json is.

Author: Changed it to:
Sample `profiles.json` content, as obtained from the Auth0 Management API:

Also fixed those headings.

              },
              "hashedPassword": {
                "value": "$2b$10$d.GvZhGwTllA7OdAmsA75uGGzqr/mhdQoU88M3zD.fX3Vb8Rcf33."
              }
            }
          }
        ]
      }
    ]
  },
  "timeout": "30m0s"
}'
```

## 6. Testing the Migration

### 6.1. Test User Login

Use the **ZITADEL login page** or your integrated app to test logging in with one of the imported users.

> **Password for the sample user:** `Password1!`

Confirm that the migrated password works as expected.

---

### 6.2. Troubleshooting

**Common issues:**

- Missing password hashes  
- Malformed JSON  
- Invalid or incomplete user data  

The import endpoint returns an `errors` array which can help you identify any issues with the import.

#### Where to check logs and get help

You can also verify that a user was imported by calling the **events endpoint** and checking for the following event type:

```json
"user.human.added"
```

## 7. Q&A and Further Resources

### Real-World Scenarios & Common Questions

**Q:** What is the maximum number of users that can be imported in a single batch?  
**A:** There is no hard limit on the number of users. However, there is a **timeout**.  

> For **ZITADEL Cloud deployments**, the timeout is **5 minutes**, which typically allows for importing around **5,000 users per batch**.

Reviewer: This can be confusing since the timeout in the http call is set to 30mins.

Can we avoid confusion here?
